File: backend-python/skills.py
import json
import os
from datetime import datetime, timezone
from typing import List, Optional
from uuid import uuid4

from models import Skill

DATA_PATH = os.path.join(os.path.dirname(__file__), "data", "skills.json")


def _read() -> List[dict]:
    if not os.path.exists(DATA_PATH):
        return []
    with open(DATA_PATH) as f:
        return json.load(f)


def _write(skills: List[dict]) -> None:
    os.makedirs(os.path.dirname(DATA_PATH), exist_ok=True)
    with open(DATA_PATH, "w") as f:
        json.dump(skills, f, indent=2)


def list_skills(query: Optional[str] = None) -> List[Skill]:
    skills = [Skill(**s) for s in _read()]
    if not query:
        return skills
    q = query.lower()
    return [
        s for s in skills
        if q in s.name.lower() or q in s.description.lower() or q in s.content.lower()
    ]


def get_skill_by_id(id: str) -> Optional[Skill]:
    for s in _read():
        if s["id"] == id:
            return Skill(**s)
    return None


def create_skill(name: str, description: str, content: str) -> Skill:
    skills = _read()
    skill = Skill(
        id=str(uuid4()),
        name=name,
        description=description,
        content=content,
        createdAt=datetime.now(timezone.utc).isoformat(),
    )
    skills.append(skill.model_dump())
    _write(skills)
    return skill


def delete_skill(id: str) -> bool:
    skills = _read()
    new_skills = [s for s in skills if s["id"] != id]
    if len(new_skills) == len(skills):
        return False
    _write(new_skills)
    return True

File: backend/src/skills.ts
import fs from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';

const DATA_PATH = path.join(__dirname, '../data/skills.json');

export interface Skill {
  id: string;
  name: string;
  description: string;
  content: string;
  createdAt: string;
}

function readSkills(): Skill[] {
  const raw = fs.readFileSync(DATA_PATH, 'utf-8');
  return JSON.parse(raw);
}

function writeSkills(skills: Skill[]): void {
  fs.writeFileSync(DATA_PATH, JSON.stringify(skills, null, 2));
}

export function listSkills(query?: string): Skill[] {
  const skills = readSkills();
  if (!query) return skills;
  const q = query.toLowerCase();
  return skills.filter(
    (s) =>
      s.name.toLowerCase().includes(q) ||
      s.description.toLowerCase().includes(q) ||
      s.content.toLowerCase().includes(q)
  );
}

export function getSkillById(id: string): Skill | undefined {
  return readSkills().find((s) => s.id === id);
}

export function createSkill(input: { name: string; description: string; content: string }): Skill {
  const skills = readSkills();
  const skill: Skill = {
    id: uuidv4(),
    ...input,
    createdAt: new Date().toISOString(),
  };
  skills.push(skill);
  writeSkills(skills);
  return skill;
}

export function deleteSkill(id: string): boolean {
  const skills = readSkills();
  const index = skills.findIndex((s) => s.id === id);
  if (index === -1) return false;
  skills.splice(index, 1);
  writeSkills(skills);
  return true;
}
